hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/OnHeapMapScanPTest.java

```diff
@@ -262,6 +290,170 @@ public void test_filterMatchingOneRow() {
+
+        HazelcastInstance instance2 = FACTORY.newHazelcastInstance(getInstanceConfig());
+
+        Thread.sleep(200L);
+        try {
+            // Await for migration stamp to change.
+            assertTrueEventually(() -> assertNotEquals(migrationStamp, mapProxy.getService().getMigrationStamp()));
+
+            // Try advance, should fail.
+            QueryException exception = assertThrows(QueryException.class, processor::complete);
+            assertEquals(SqlErrorCode.PARTITION_DISTRIBUTION, exception.getCode());
+            assertTrue(exception.isInvalidatePlan());
+        } finally {
+            instance2.shutdown();
+        }
+    }
+
+    @Test
+    public void testConcurrentMapDestroy() {
+        HazelcastInstance instance1 = instance().getHazelcastInstance();
+        IMap<TestKey, TestValue> map = instance1.getMap(MAP_OBJECT);
+        List<Object[]> expected = new ArrayList<>();
+
+        // Collect local partitions.
+        PartitionIdSet partitionIdSet = new PartitionIdSet(PARTITION_COUNT);
+
+        for (int i = 0; i < PARTITION_COUNT; i++) {
+            for (Partition partition : instance1.getPartitionService().getPartitions()) {
+                if (instance1.getLocalEndpoint().getUuid().equals(partition.getOwner().getUuid())) {
+                    partitionIdSet.add(partition.getPartitionId());
+                    break;
+                }
+            }
+        }
+
+        for (int partition : partitionIdSet) {
+            int currentKey = 0;
+
+            while (true) {
+                TestKey key = new TestKey(currentKey);
+
+                if (instance1.getPartitionService().getPartition(key).getPartitionId() == partition) {
+                    map.put(key, new TestValue(1, true));
+                    expected.add(new Object[]{1});
+                    break;
+                }
+                currentKey++;
+            }
+        }
+
+        MapScanMetadata scanMetadata = new MapScanMetadata(
+                map.getName(),
+                GenericQueryTargetDescriptor.DEFAULT,
+                GenericQueryTargetDescriptor.DEFAULT,
+                Collections.singletonList(valuePath("val2")),
+                Collections.singletonList(INT),
+                singletonList(ColumnExpression.create(0, INT)),
+                new ConstantPredicateExpression(true)
+        );
+
+        TestSupport
+                .verifyProcessor(adaptSupplier(OnHeapMapScanP.onHeapMapScanP(scanMetadata)))
+                .hazelcastInstance(instance1)
+                .jobConfig(new JobConfig().setArgument(SQL_ARGUMENTS_KEY_NAME, emptyList()))
+                .outputChecker(LENIENT_SAME_ITEMS_ANY_ORDER)
+                .disableSnapshots()
+                .disableProgressAssertion()
+                .expectOutput(expected);
+
+        // Destroy the map.
+        instance1.getMap(MAP_OBJECT).destroy();
+        try {
+            Thread.sleep(50L);
```

Comment:
I would not even try to rely on `sleep()` here. We need better way to ensure `destroy()` is concurrent to map scan.

Reply:
Any suggestions? For me, it's the simplest way to wait until map will be destroyed.

Reply:
But it does not guarantee that the map will be destroyed before the verification (I understand this is what you are trying to achieve here) + it slows down the test. What about `MapServiceContext.getMapContainers().keySet()`?